### PR TITLE
[core][lua] Physical Mobskill Refactor

### PR DESCRIPTION
**Pull request by UmeboshiXI**
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I understand I should leave resolving conversations to the LandSandBoat team so that reviewers won't miss what was said.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## What does this pull request do?

Submitting as draft for now. Want to bounce it off automated checks while I sleep and see where it stands/get eyes on it. Will return to fix/consolidate things in morning.

Aware of debug prints being present, will remove/comment out before final undrafted PR but leaving in at the moment for testing purposes.

Did extensive testing on my local but will need to retest based on further changes/adjustments so leaving that box unchecked for now.

### 1. Reworks how shadow consumption is handled.
 - Damage is no longer influenced by the shadows consumed.
 
 - Based on retail testing, shadow consumption for mobskills was found to not be affected by hit rate checks.
 
 - Based on retail testing, shadows are not calculated per hit. A single hit skill that takes 3 shadows will always take 3 shadows.
 
### 2. Reworked AoE skill shadow mitigation mechanic based on retail testing.
 - Physical AoE skills that do not ignore/wipe shadows have a chance to not consume the normal shadow amount.
 
 - The amount of shadow consumption mitigated is based on how many shadows a skill would normally take.
 
 - The amount of shadows a skill would normally take acts as a counter for how many mitigation attempts will be made. Example: a skill would take 4 shadows will have 4 attempts made for mitigation.
 
 - Mitigation can not reduce shadow consumption to 0. 1 shadow will always be taken at a minimum.
 
 - Test data can be found [here](https://docs.google.com/spreadsheets/d/12O11XhwBp18nA7XEMx-MIYPxxdlwpXkWxMIIBzo1QqU/edit?gid=1541834103#gid=1541834103) and [here](https://docs.google.com/spreadsheets/d/12O11XhwBp18nA7XEMx-MIYPxxdlwpXkWxMIIBzo1QqU/edit?gid=379130688#gid=379130688).
 
###  3. Refactors Physical/Ranged mobskills
 - Adds/extends support for various parameters.
 
  - Retires obsolete xi.mobskills.mobPhysicalHit() function. Superseded by xi.mobskills.processDamage() which checks if the skill's hits landed.
 
 - takeDamage() is now wrapped with xi.mobskills.processDamage() to prevent shadow chip damage. Status Effect applications are also gated in the same way. Closes related issue: https://github.com/LandSandBoat/server/issues/8880
 
 - Retires obsolete xi.mobskills.mobPhysicalStatusEffectMove() function. Superseded with usage pattern of mobStatusEffectMove() wrapped with xi.mobskills.processDamage().
 
 - Retires obsolete mobPhysicalDrainMove() function. Superseded with usage pattern of mobDrainMove() wrapped with xi.mobskills.processDamage().
 
 - Physical skills now calculate individual hits and record hit information that can be called later.
 
 - Ranged skills now use ranged PDIF, FSTR, and Crit functions.
 
 - Fixes ranged skill TP Gain by adjusting the getBaseRangedDelay() for mobs. Based on retail testing, mob ranged delay seems to be the same for all normal mobs recorded. Test Data can be found [here](https://docs.google.com/spreadsheets/d/12O11XhwBp18nA7XEMx-MIYPxxdlwpXkWxMIIBzo1QqU/edit?gid=1320693318#gid=1320693318).
 
 - Added missing trySkillUp() checks for Evasion. This is calculated per hit based on retail.
 
 - Adjusted Shield Mastery checks for shield blocks. We previously gated this behind having the Shield Mastery job trait but through retail testing, it was found that it could still proc if you had equipment that had Shield Mastery mods.
 
 - Implements skill data from Jimmayus's spreadsheet where applicable. Recent PRs that had updated capture data were kept/took priority.
 
 - Added mob families/notes to each skill's script for easy reference.
 
 - Added TODO: notes to mobskill scripts where we need capture data or code adjustments.
 
## Steps to test these changes
Tested these steps on my local and were working then. Will retest again if I make further edits.

1. Test/play around with parameters. Nothing should error out. Each one should affect its related stat as expected.

2. Test shadow consumption.
 - Single hit skills should simply take the amount of shadows stated in the skill's script. If the player has enough shadows to block the skill, it gets absorbed. If not, they take full damage.
  - AOE/Conal physical/ranged skills that do not ignore/wipe shadows can have varying shadows taken due to mitigation mechanics. A skill that would take 4 shadows can range from 1-4 shadows taken.
  - Shadows are calculated before hitrate checks and are not calculated per hit. This means the function is ran once and is an all or nothing affair.
  
 3. Test Third Eye.
  - Third Eye now blocks single target multi hit skills. If it procs, it will exit the attack loop in event of multi hits.
  - Third Eye does not block and is wiped by AOE mob skills.
  - Third Eye does not interact with Magical/Breath mob skills.
  
 4. Test ranged skill blocks/parries/guards
  - Currently, ranged skills should not be able to be parried/blocked/guarded. The appropriate params are set to skip these functions in the skill scripts scripts. Left the possibility of toggling it on in event we find edge cases.
  
 5. Test skill messaging for skill attacks.
  - param.primaryMessage can be used to override the default damage message. This is mostly used for skill attacks.
  - Supports appropriate corresponding miss message based on the primaryMessage type.

### CODE REVIEW

File: scripts/utils/combat_utils.lua [hunk 1]
```diff
@@ -104,6 +124,70 @@ function utils.takeShadows(actor, damage, shadowsToRemove)
...
+        end
+
+        -- Target had enough shadows to block.
+        return true, actualConsumed
+    end
+
+    -- Fallback if we somehow got this far.
+    return false, shadowsToRemove
+end
+
```
Comment (Xaver-DaRed) [inline scripts/utils/combat_utils.lua:173]: I see a few issues with current implementation. Most notably, how we delete both blink and utsusemi. Blink may have not triggered and hence shouldnt be consumed.

Proposal:
```lua
function utils.shadowAbsorb(target, shadowsToRemove)
    local utsusemiMod = target:getMod(xi.mod.UTSUSEMI)
    local blinkMod    = target:getMod(xi.mod.BLINK)

    -- Early return: Target has no shadows.
    if
        utsusemiMod == 0 and
        blinkMod == 0
    then
        return false, 0
    end

    local targetShadows = 0

    -- Utsusemi takes precedence over blink.
    if utsusemiMod > 0 then
        targetShadows = utsusemiMod

        local effect = target:getStatusEffect(xi.effect.COPY_IMAGE)
        if effect then
            if targetShadows == 0 then
                target:delStatusEffect(xi.effect.COPY_IMAGE)
            elseif targetShadows == 1 then
                effect:setIcon(xi.effect.COPY_IMAGE)
            elseif targetShadows == 2 then
                effect:setIcon(xi.effect.COPY_IMAGE_2)
            else
                effect:setIcon(xi.effect.COPY_IMAGE_3)
            end
        end

        target:setMod(xi.mod.UTSUSEMI, targetShadows)

    -- Blink has a random chance of triggering when no utsusemi is present.
    elseif blinkMod > 0 then
        if math.random(1, 100) <= 20 then
            return false, 0
        end

        targetShadows = blinkMod

        if targetShadows == 0 then
            target:delStatusEffect(xi.effect.BLINK)
        end

        target:setMod(xi.mod.BLINK, targetShadows)
    end

    local actualConsumed = utils.clamp(shadowsToRemove, 0, targetShadows)
    local absorbHit      = targetShadows >= shadowsToRemove

    return absorbHit, actualConsumed
end
```

Comment (WinterSolstice8) [reply scripts/utils/combat_utils.lua:173]: As far as I know those effects are mutually exclusive. You can either have blink or utsusemi

File: scripts/globals/combat/physical_utilities.lua [hunk 2]
```diff
@@ -852,6 +852,17 @@ xi.combat.physical.criticalRateFromStatDiff = function(actor, target)
...
 
+-- dStat: Ranged critical hit rate bonus from AGI vs AGI difference.
+xi.combat.physical.criticalRateFromAGIDiff = function(actor, target)
+    local statBonus = 0
+
+    local dAgi = math.max(0, actor:getStat(xi.mod.AGI) - target:getStat(xi.mod.AGI))
+    statBonus = math.floor(dAgi / 10)
+    statBonus = statBonus / 100
+
+    return statBonus
```
Comment (WinterSolstice8) [inline scripts/globals/combat/physical_utilities.lua:863]: Surely there's a cap... interesting

File: scripts/globals/mobskills.lua [hunk 3]
```diff
@@ -97,245 +104,997 @@ local function calculateMobMagicBurst(target, actionElement, skillchainCount)
...
-    if tp >= 1000 and tp < 1500 then
-        return ftp1 + (((ftp2 - ftp1) / 500) * (tp - 1000))
-    elseif tp >= 1500 and tp <= 3000 then
-        -- generate a straight line between ftp2 and ftp3 and find point @ tp
-        return ftp2 + (((ftp3 - ftp2) / 1500) * (tp - 1500))
-    end
-
-    return 1 -- no ftp mod
+-- Helper function to store default hit information before being modified.
+local function defaultHitInfo(hitNumber)
```
Comment (WinterSolstice8) [inline scripts/globals/mobskills.lua:205]: We should probably have LLS type hinting for this table that it returns and also make `handleSinglePhysicalHit` and `defaultHitInfo`, `handleSingleRangedHit`, etc return that value

iirc it's something like
`---@alias something { hitNumber: number, hitLanded: boolean }`
then
`---@return something`

File: scripts/globals/mobskills.lua [hunk 4]
```diff
@@ -97,245 +104,997 @@ local function calculateMobMagicBurst(target, actionElement, skillchainCount)
...
+        if guaranteedFirstHit and hitNumber == 1 then
+            hitChance = 1
+        elseif hitNumber == 1 then
+            -- First hit gets bonus accuracy. TODO: Confirm this and the amount if true.
+            hitChance = xi.combat.physicalHitRate.getRangedHitRate(mob, target, 100 + accuracyModifier + targetSpecialAttackEvasion, false)
+        else
+            hitChance = xi.combat.physicalHitRate.getRangedHitRate(mob, target, accuracyModifier + targetSpecialAttackEvasion, false)
+        end
+
+        print('hitChance', hitChance)
```
Comment (WinterSolstice8) [inline scripts/globals/mobskills.lua]: leftover print

File: scripts/globals/mobskills.lua [hunk 5]
```diff
@@ -97,245 +104,997 @@ local function calculateMobMagicBurst(target, actionElement, skillchainCount)
...
+    -- Initialize return structure
+    returnInfo.damage           = 0
+    returnInfo.hybridDamage     = 0
+    returnInfo.hitsLanded       = 0
+    returnInfo.attackType       = attackType
+    returnInfo.damageType       = damageType
+    returnInfo.hybridAttackType = hybridAttackType
+    returnInfo.hybridDamageType = hybridDamageType
+    returnInfo.isCritical       = false
+    returnInfo.hitData          = {}
```
Comment (Frankie-hz) [inline scripts/globals/mobskills.lua]: I see these default values are used in both ranged and physical skills but you define them differently.

Example: You have `utils.defaultIfNil(skillParams.skipStoneskin, false)` in here but in the other one you have `utils.defaultIfNil(skillParams.skipStoneskin and true, false)`

Since these appear to the be the exact same you can make a local helper function to return all these values which will lower the complexity and remove duplicated code. I am not sure if these need to be different in the future but as its proposed they appear to be identical

File: scripts/globals/combat/physical_utilities.lua (hunk 2 above)
Comment (UmeboshiXI) [reply scripts/globals/combat/physical_utilities.lua:863]: I pulled this from our current implementation on core side.

GetAGICritBonus() in battleutils.cpp.

https://www.bg-wiki.com/bg/Critical_Hit_Rate
A comment in that function linking to bgwiki mentions there is no known cap to dAGI. Would need to do tests to confirm.

File: src/map/entities/battleentity.cpp [hunk 6]
```diff
@@ -2827,7 +2827,7 @@ void CBattleEntity::OnMobSkillFinished(CMobSkillState& state, action_t& action)
         if (PSkill->hasMissMsg())
         {
             result.resolution = ActionResolution::Miss;
-            result.param      = 0;
+            result.param      = damage;
```
Comment (sruon) [inline src/map/entities/battleentity.cpp]: What did you attempt to fix with this change? Miss action packets always have param set to 0.

This is what's causing your tests to fail.

File: scripts/utils/combat_utils.lua (hunk 1 above)
Comment (UmeboshiXI) [reply scripts/utils/combat_utils.lua:173]: Cleaned up the code a bit but Blink and Utsusemi are exclusive and can not be on at the same time.

File: scripts/globals/mobskills.lua (hunk 4 above)
Comment (UmeboshiXI) [reply scripts/globals/mobskills.lua]: Removed all my debug prints as of now.

File: src/map/entities/battleentity.cpp (hunk 6 above)
Comment (UmeboshiXI) [reply src/map/entities/battleentity.cpp]: Test passed after implementing the fix in CBattleEntity::OnMobSkillFinished()

**Review by Xaver-DaRed — APPROVED** [review]

File: scripts/globals/mobskills.lua [hunk 7]
```diff
@@ -97,245 +108,884 @@ local function calculateMobMagicBurst(target, actionElement, skillchainCount)
...
-    if hitslanded == 0 then
-        finaldmg   = 0
-        hitslanded = 0
-        skill:setMsg(xi.msg.basic.SKILL_MISS)
-    -- calculate tp return of mob skill and add if hit primary target
-    elseif skill:getPrimaryTargetID() == target:getID() and finaldmg > 0 then
-        local tpReturn = xi.combat.tp.getSingleMeleeHitTPReturn(mob, false)
-        tpReturn = tpReturn + 10 * (hitslanded - 1) -- extra hits give 10 TP each
-        mob:addTP(tpReturn)
+    -- TODO: Need more research on hybrid skills. Is the magical damage strictly reliant on the physical hit doing damage?
```
Comment (WinterSolstice8) [inline scripts/globals/mobskills.lua:964]: you can probably just assume it works like summoner BP hybrids (which is probably the same as players?) But that said I'm not aware of a hybrid mobskill
